subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/h2/H2PersistentMap.java

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cache.internal.h2;
+
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
+import org.gradle.cache.internal.btree.PersistentMap;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder;
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder;
+import org.h2.Driver;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.sql.Blob;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class H2PersistentMap<V> implements PersistentMap<String, V> {
+
+    private final HikariDataSource dataSource;
+    private Serializer<V> serializer;
+
+    public H2PersistentMap(Serializer<V> serializer, Path dbPath, int maxPoolSize) {
+        this.serializer = serializer;
+        this.dataSource = createHikariDataSource(dbPath, maxPoolSize);
+    }
+
+    private static HikariDataSource createHikariDataSource(Path dbPath, int maxPoolSize) {
+        HikariConfig hikariConfig = new HikariConfig();
+        // RETENTION_TIME=0 prevents uncontrolled DB growth with old pages retention
+        // We use MODE=MySQL so we can use INSERT IGNORE
+        String h2JdbcUrl = String.format("jdbc:h2:file:%s;RETENTION_TIME=0;MODE=MySQL;INIT=runscript from 'classpath:/h2/schemas/org.gradle.caching.local.internal.h2.H2PersistentMap" +
+            ".sql'", dbPath.resolve("store"));
+        hikariConfig.setJdbcUrl(h2JdbcUrl);
+        hikariConfig.setDriverClassName(Driver.class.getName());
+        hikariConfig.setUsername("sa");
+        hikariConfig.setPassword("");
+        hikariConfig.setCatalog("store");
+        hikariConfig.setPoolName("store-pool");
+        hikariConfig.setMaximumPoolSize(maxPoolSize);
+        hikariConfig.setConnectionInitSql("select 1;");
+        return new HikariDataSource(hikariConfig);
+    }
+
+    @Override
+    public V get(String key) {
+        try (Connection conn = dataSource.getConnection()) {
+            try (PreparedStatement stmt = conn.prepareStatement("select entry_content from store.catalog where entry_key = ?")) {
+                stmt.setString(1, key);
+                try (ResultSet rs = stmt.executeQuery()) {
+                    if (rs.next()) {
+                        Blob content = rs.getBlob(1);
+                        try (InputStream binaryStream = content.getBinaryStream()) {
+                            return serializer.read(new KryoBackedDecoder(binaryStream));
+                        }
+                    }
+                    return null;
+                }
+            }
+        } catch (Exception e) {
```

Comment:
❕I think the get/set would be better by splitting retrieval/insertion of the `Blob` from the DB, and deserializing/serializing it into separate methods.  

Then you could catch only `SqlException` for the DB part, and `IOException` for the serialization part, in addition to having methods with single responsibilities.